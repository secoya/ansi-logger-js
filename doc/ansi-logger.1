.\" Manpage for ansi-logger.
.\" Contact briankchritensen@gmail.com to correct errors or typos.

.TH man 1 "09 Nov 2017" "3.1.7" "ansi-logger(1) man page"

.SH NAME
ansi-logger \- format JSON entries to colored output.

.SH SYNOPSIS
.B ansi-logger
[\fB\-s\fR]
[\fB\-l\fR \fILEVEL\fR | \fB-m\fR \fIMASKS\fR]
[\fB\-t\fR \fISECONDS\fR]
[\fB\-f\fR \fIFILE\fR]
[\fB\-\-mapper\fR \fIFILE_OR_STRING\fR]

.SH DESCRIPTION
.RE
Creating logs with the JSONTransformer of the ansi-logger library,
.RE
is great for production environment for forwarding to log services like logstash.
.RE
This tool can parse the JSON logs and output it via the TextTransformer.

.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Show this help message.

.TP
.BR \-f ", " \-\-file =\fIFILE\fR
The log file to parse and format [default: -].

.TP
.BR \-l ", " \-\-loglevel =\fILEVEL\fR
The output levle [default: VERBOSE].

.TP
.BR \-m ", " \-\-logmasks =\fIMASKS\fR
Comma separated list of log masks to output.

.TP
.BR \-s ", " \-\-split-pipes
Direct error entries to stderr.

.TP
.BR \-t ", " \-\-timeout =\fISECONDS\fR
The timeout for receiving data [default: 1]. 0 for disable.

.TP
.BR \-\-mapper =\fIFILE_OR_STRING\fR
JavaScript file that exports a single function to map parsed JSON structure to LogEntry
.br
or a comma separated mapper pairs, e.g \fIlevelText\fR=\fBlevel\fR,\fIgroup\fR=\fBkey\fR
.br
.LP
.RS 7

Input:
.br
{"timestamp": "2017-11-08 10:19:01+01:00", "level": "info", "import_key": "references", "glob": "*.xml", "message": "Importing files"}

// mapper.js
.br
module.exports = entry => {
    let levelText;
    switch (entry.level) {
        case 'fatal':
            levelText = 'ERROR';
            break;
        default:
            levelText = typeof entry.level === 'string' ? entry.level.toUpperCase() : 'LOG';
    }

    return {
        group: entry.import_key,
        levelText: levelText,
        message: entry.message,
        timestamp: entry.timestamp,
    };
.br
};

Output:
.br
[2017-11-08 10:19:01+01:00] [references] [INFO] Importing files

.RE

.SH LEVEL OPTIONS
.TP
.IR SILENT
Silent all output.

.TP
.IR ERROR
Output only error entries.

.TP
.IR WARN
Output error and warn entries.

.TP
.IR LOG
Output error, warn and log entries.

.TP
.IR INFO
Output error, warn, log and info entries.

.TP
.IR DEBUG
Output error, warn, log, info and debug entries.

.TP
.IR VERBOSE
Output error, warn, log, info, debug and verbose entries.

.SH MASKS OPTIONS

.TP
.IR ERROR
Output error entries.

.TP
.IR WARN
Output warn entries.

.TP
.IR LOG
Output log entries.

.TP
.IR INFO
Output info entries.

.TP
.IR DEBUG
Output debug entries.

.TP
.IR VERBOSE
Output verbose entries.

.SH EXAMPLES
.TP
Only output log entries from INFO and DEBUG
tail -n200 -f big.log | \fBansi-logger\fR \fB\-m\fR \fIINFO\fR,\fIDEBUG\fR

.TP
Only output log entries from ERROR and WARN
tail -n200 -f big.log | \fBansi-logger\fR \fB\-l\fR \fIWARN\fR

.TP
Redirect file to stdin
\fBansi-logger\fR < ./big.log | less

.TP
Use file switch to read a log file
\fBansi-logger\fR \fB\-f\fR \fI./big.log\fR | less

.TP
Use grep combined with \fBtail -f\fR
\fBtail -f\fR ./big.log | grep \fI--line-buffered\fR 'UserID: 4' | \fBansi-logger\fR -t10 -mWARN,DEBUG

.SH SEE ALSO
\fBansi-logger\fR(3)

.SH BUGS
Report bugs at https://github.com/secoya/ansi-logger-js/issues

.SH AUTHOR
Brian K. Christensen <briankchritensen@gmail.com>
