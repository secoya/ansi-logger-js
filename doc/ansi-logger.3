.\" Manpage for ansi-logger library.
.\" Contact briankchristensen@gmail.com to correct errors or typos.

.TH man 3 "10 Nov 2017" "3.2.0" "ansi-logger(3) man page"

.SH NAME
ansi-logger \- javascript logging library.

.SH SYNOPSYS

.TP
.B createTextLogger(\fIoptions\fR?:\ \fBPartial<Options>)\fR:\ \fIAnsiLogger<TextTransformer>
.RS 7
Create a simple text logger, generally used to output human readable
format to the console.
.RE

.TP
.BR createJSONLogger(\fIoptions\fR?:\ \fBPartial<Options>)\fR:\ \fIAnsiLogger<JSONTransformer>
.RS 7
Create simple JSON logger, generally used for when outputting
to log services like e.g. Log Stash.
.RE

.TP
.B createLoggerFromEnvironment(\fIoptions\fR?:\ \fBPartial<Options>)\fR:\ \fIAnsiLogger<Transformer<string>>
.RS 7
Create logger based on the options passed, but let environment variables overwrite the options if they are defeined.
.br
Default is JSON logger with log level INFO.
.LP
The environment variables are:
.br
- \fBLOGLEVEL\fR   possible values are:
.br
.RS 13
\fISILENT\fR | \fIERROR\fR | \fIWARN\fR | \fISUCCESS\fR | \fILOG\fR | \fIINFO\fR | \fIDEBUG\fR | \fIVERBOSE\fR | \fInumber\fR
.RE
.br
- \fBLOGFORMAT\fR  possible values are:
.br
.RS 13
\fITEXT\fR | \fIJSON\fR
.RE
.RE

.TP
.BR AnsiLogger<\fIE\fB>(\fIoptions\fR:\ \fBPartial<Options>)
Set options changes the internal setting of the logger and the behavior will change after this call.
.br
If you change the log levels to output the new mask of log levels is outputted to the INFO level,
.br
so if INFO level isn't in the new log level mask, nothing will be outputted.
.RS 7
.TP
.BR error<\fIL\fB>(\fImsg\fR:\ \fIL\fB,\ ...rest:\ any[])\fR:\ \fIL
Print message(s) to error log level if enabled.

.TP
.BR log<\fIL\fB>(\fImsg\fR:\ \fIL\fB,\ ...rest:\ any[])\fR:\ \fIL
Print messeage(s) to default log level if enabled.

.TP
.BR success<\fIL\fB>(\fImsg\fR:\ \fIL\fB,\ ...rest:\ any[])\fR:\ \fIL
Print a success formatted message.

.TP
.BR warn<\fIL\fB>(\fImsg\fR:\ \fIL\fB,\ ...rest:\ any[])\fR:\ \fIL
Print a warning formatted message.

.TP
.BR debug<\fIL\fB>(\fImsg\fR:\ \fIL\fB,\ ...rest:\ any[])\fR:\ \fIL
Print message(s) to debug log level if enabled.

.TP
.BR verbose<\fIL\fB>(\fImsg\fR:\ \fIL\fB,\ ...rest:\ any[])\fR:\ \fIL
Print a verbose formatted message.

.TP
.BR formatTypes<\fIL\fB>(\fImsg\fR:\ \fIL\fB)\fR:\ \fIL
Text transform Format types to string, some types make resively calls.

.TP
.BR formatError(\fIerr\fR:\ \fBError)\fR:\ \fIvoid
Format an error. This is typically used, for text transforming Exception(s)/Error(s).
.RE

.TP
.BR JSONTransformer(\fIoptions\fR:\ \fBPartial<Options>)
Transform log entries JSON encoded string output.

.RS 7

.TP
.BR format(\fIentry\fR:\ \fBLogEntry)\fR:\ \fIstring
Format log entries to json strings.

.TP
.BR formatTypes<\fIT\fB>(\fIinputType\fR:\ \fIT\fB)\fR:\ \fIT
Don't "format" complex types, just pass them along,
.br
and let the formatting happen in the `format` method.
.TP
.BR printer.err(\fImsg\fR:\ \fBstring)\fR:\ \fIvoid
Print transformed error log entries to output stream.

.TP
.BR printer.out(\fImsg\fR:\ \fBstring)\fR:\ \fIvoid
Print transformed log entries to output stream.

.RE

.TP
.BR TextTransformer(\fIoptions\fR:\ \fBPartial<Options>)
Transformer from log entries to human readable colorized text.

.RS 7

.TP
.BR format(\fIentry\fR:\ \fBLogEntry):\ \fIstring
Transform log entry to text output.

.TP
.BR formatTypes<\fIT\fB>(\fIinputType\fR:\ \fIT\fB)\fR:\ \fIT
Format complex types.

.TP
.BR printer.err(\fImsg\fR:\ \fBstring):\ \fIvoid
Print transformed error log entries to output stream.

.TP
.BR printer.out(\fImsg\fR:\ \fBstring):\ \fIvoid
Print transformed log entries to output stream.

.RE
.TP
.BR IdentityTransformer(\fIoptions\fR:\ \fBPartial<Options>)
The identity transformer makes no transformation to the log entry
.br
and just returns the `LogEntry` as is.

.RS 7

.TP
.BR format(\fIentry\fR:\ \fBLogEntry):\ \fIstring
Format log entry according to identity rules.

.TP
.BR formatTypes<\fIT\fB>(\fIinputType\fR:\ \fIT\fB)\fR:\ \fIT
Return the complex type identity.

.TP
.BR printer.err(\fImsg\fR:\ \fBLogEntry)\fR:\ \fIvoid
Print transformed error log entries to output stream.

.TP
.BR printer.out(\fImsg\fR:\ \fBLogEntry)\fR:\ \fIvoid
Print transformed log entries to output stream.

.RE

.SH OPTIONS

.SS createTextLogger

.TP
.BR \fIgroup\fR:\ string
The group to mark the log entries with.

.TP
.BR \fIgroupColor\fR:\ (str:\ string)\ =>\ string
The color to show the group in.

.TP
.BR \fIlogLevel\fR:\ number
The log level mask.

.SS createJSONLogger

.TP
.BR \fIgroup\fR:\ string
The group to mark the log entries with.

.TP
.BR \fIlogLevel\fR:\ number
The log level mask.

.SS createLoggerFromEnvironment

.TP
.BR \fIgroup\fR:\ string
The group to mark the log entries with.

.TP
.BR \fIgroupColor\fR:\ (str:\ string)\ =>\ string
The color to show the group in.

.TP
.BR \fIlogFormat\fR:\ 'TEXT'\ |\ 'JSON'
The format to use if \fBLOGFORMAT\fR is not defined.

.TP
.BR \fIlogLevel\fR:\ number
The log level mask.

.SS AnsiLogger

.TP
.BR \fIgroup\fR:\ string
The group to mark the log entries with.

.TP
.BR \fIlogLevel\fR:\ number
The log level mask.

.TP
.BR \fItimeformat\fR:\ string
Moment.js formats.

.TP
.BR \fItransformer\fR:\ Transformer
The transformer to convert LogEntry types to what the outputters can print.

.SS JSONTranformer

.TP
.BR \fIprinter.err\fR:\ (msg:\ string)\ =>\ void
Print transformed error log entries to output stream.

.TP
.BR \fIprinter.out\fR:\ (msg:\ string)\ =>\ void
Print transformed log entries to output stream.

.SS TextTranformer

.TP
.BR \fIcolorMap\fR:\ Partial<ColorMap>
The map of which coloring functions to use when/where.

.TP
.BR \fIcolors\fR:\ boolean
Whether or not if colors is enabled, default: process.stdout.isTTY.

.TP
.BR \fIforceColors\fR:\ boolean
Always output colors, no matter if stdout is a TTY.

.TP
.BR \fIprinter.err\fR:\ (msg:\ string)\ =>\ void
Print transformed error log entries to output stream.

.TP
.BR \fIprinter.out\fR:\ (msg:\ string)\ =>\ void
Print transformed log entries to output stream.

.SS IdentityTransformer

.TP
.BR \fIprinter.err\fR:\ (msg:\ string)\ =>\ void
Print transformed error log entries to output stream.

.TP
.BR \fIprinter.out\fR:\ (msg:\ string)\ =>\ void
Print transformed log entries to output stream.

.SS Mask
The numeric level for marking the log entries.

.TP
.BR \fIERROR\fR\ \ \ \ (1)
The mask for error log entries.

.TP
.BR \fIWARN\fR\ \ \ \ \ (2)
The mask for warning log entries.

.TP
.BR \fISUCCESS\fR\ \ (4)
The mask for success log entries.

.TP
.BR \fILOG\fR\ \ \ \ \ \ (8)
The mask for default log entries.

.TP
.BR \fIINFO\fR\ \ \ \ (16)
The mask for info log entries.

.TP
.BR \fIDEBUG\fR\ \ \ (32)
The mask for debug log entries.

.TP
.BR \fIVERBOSE\fR\ (64)
The mask for verbose log entries.

.SS Level
The numeric level for marking the log entries.

.TP
.BR \fISILENT\fR\ \ \ \ (0)
This log level won't output anything.

.TP
.BR \fIERROR\fR\ \ \ \ \ (1)
This log level only outputs error log entries.

.TP
.BR \fIWARN\fR\ \ \ \ \ \ (3)
This log level outputs both warning and error log entries.

.TP
.BR \fISUCCESS\fR\ \ \ (7)
This log level outputs all both inclusive the success log entries.

.TP
.BR \fILOG\fR\ \ \ \ \ \ (15)
This log level outputs all both inclusive the default log entries.

.TP
.BR \fIINFO\fR\ \ \ \ \ (31)
This log level outputs all both inclusive the info log entries.

.TP
.BR \fIDEBUG\fR\ \ \ \ (63)
This log level outputs all both inclusive the debug log entries.

.TP
.BR \fIVERBOSE\fR\ (127)
This log level outputs everything.

.SS ColorMap

\fIERROR\fR:\ \ \ (str:\ string)\ =>\ string
.br
\fIWARN\fR:\ \ \ \ (str:\ string)\ =>\ string
.br
\fISUCCESS\fR:\ (str:\ string)\ =>\ string
.br
\fILOG\fR:\ \ \ \ \ (str:\ string)\ =>\ string
.br
\fIINFO\fR:\ \ \ \ (str:\ string)\ =>\ string
.br
\fIDEBUG\fR:\ \ \ (str:\ string)\ =>\ string
.br
\fIVERBOSE\fR:\ (str:\ string)\ =>\ string
.br
\fITIME\fR:\ \ \ \ (str:\ string)\ =>\ string
.br
\fIGROUP\fR:\ \ \ (str:\ string)\ =>\ string

.SH ENVIRONMENT
\fBLOGFORMAT\fR
\fBLOGLEVEL\fR

.SH SEE ALSO
\fBansi-logger\fR(1)

.SH BUGS
Report bugs at https://github.com/secoya/ansi-logger-js/issues

.SH AUTHOR
Brian K. Christensen <briankchritensen@gmail.com>
