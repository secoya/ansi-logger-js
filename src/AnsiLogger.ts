import * as clc from 'cli-color';
import * as moment from 'moment';
import { IdentityTransformer } from './IdentityTransformer';
import { JSONTransformer } from './JSONTransformer';
import { TextTransformer } from './TextTransformer';

export interface LogEntry {
	group?: string;
	levelNumeric: number;
	levelText: string | null;
	message: string | null;
	timestamp: string;
}

export type Transformer = (this: AnsiLogger, entry: LogEntry) => any;

function assertNever(__: never, msg?: string) {
	throw new Error(msg || 'Unsupported option ' + __);
}

// tslint:disable:no-bitwise
export enum Mask {
	ERROR = 0b0000001,
	WARN = ERROR << 1,
	SUCCESS = WARN << 1,
	LOG = SUCCESS << 1,
	INFO = LOG << 1,
	DEBUG = INFO << 1,
	VERBOSE = DEBUG << 1,
}

export enum Level {
	SILENT = 0b0000000,
	ERROR = SILENT | Mask.ERROR,
	WARN = ERROR | Mask.WARN,
	SUCCESS = WARN | Mask.SUCCESS,
	LOG = SUCCESS | Mask.LOG,
	INFO = LOG | Mask.INFO,
	DEBUG = INFO | Mask.DEBUG,
	VERBOSE = DEBUG | Mask.VERBOSE,
}
// tslint:enable:no-bitwise

export interface LoggerOptionsInput {
	colors: { [color: string]: clc.Format };
	group: string;
	'group-color': clc.Format;
	'log-level': number;
	'no-colors': boolean;
	outputters: {
		err: (msg: any) => void;
		out: (msg: any) => void;
	};
	'startup-info': boolean;
	timeformat: string;
	transformer: (this: AnsiLogger, entry: LogEntry) => any;
}

export interface LoggerOptions {
	group?: LoggerOptionsInput['group'];
	'group-color'?: LoggerOptionsInput['group-color'];

	'log-level': LoggerOptionsInput['log-level'];
	'no-colors': LoggerOptionsInput['no-colors'];
	outputters: LoggerOptionsInput['outputters'];
	'startup-info': LoggerOptionsInput['startup-info'];
	timeformat: LoggerOptionsInput['timeformat'];
	transformer: LoggerOptionsInput['transformer'];
}

export type LogMask = keyof typeof Mask;
export type LogLevel = keyof typeof Level;

export type ColorMap = { [mask: number]: clc.Format } & { TIME: clc.Format } & { TITLE: clc.Format };

export function matchMask(level: number, mask: number): boolean {
	// tslint:disable-next-line:no-bitwise
	return (level & mask) === mask;
}

const knownColors = Object.keys(Level).concat(['TIME', 'TITLE']);
function isKnownColorType(color: string): color is LogLevel | 'TIME' | 'TITLE' {
	return knownColors.includes(color);
}

/**
 * Ansi output logger.
 * This controls what should be ouputted to the console,
 * everything is categorized into log levels, so when you set a log level
 * you output from all the selected levels.
 * It is possible to disables colors (some teminals don't support colors).
 * you can also specify that you are only interested in output for a specific
 * log-level, then everything else is not outputted.
 * It is also possible to make the logger silent.
 */
export class AnsiLogger {
	// tslint:enable:react-aware-member-ordering

	public get options(): Partial<LoggerOptions> {
		return this._options;
	}

	/**
	 * The options object holder.
	 * This is filled with the default values when the Logger is constructed,
	 * it can be changed by using the setOptions method.
	 */
	// tslint:disable-next-line:variable-name react-aware-member-ordering
	private _options: LoggerOptions;
	/**
	 * Default color scheme
	 * can be overriden by providing a
	 * new full or partial ColorMap.
	 */
	public readonly colors: ColorMap = {
		[Mask.ERROR]: clc.bgRed.white,
		[Mask.WARN]: clc.red.bold,
		[Mask.SUCCESS]: clc.green,
		[Mask.LOG]: clc,
		[Mask.INFO]: clc.blue,
		[Mask.DEBUG]: clc.yellow,
		[Mask.VERBOSE]: clc.magenta,
		TIME: clc.cyan,
		TITLE: clc.cyan,
	};

	/**
	 * Constructs a Logger, and sets default option values.
	 */
	public constructor(options?: Partial<LoggerOptionsInput>) {
		// tslint:disable:object-literal-key-quotes
		this._options = {
			group: undefined,
			'group-color': undefined,
			'log-level': Level.INFO,
			'no-colors': false,
			outputters: {
				out(msg: any): void {
					// Setting up default group // the log level // disbles colors if true
					process.stdout.write(msg + '\n');
				},
				err(msg: any): void {
					process.stderr.write(msg + '\n');
				},
			},
			'startup-info': true,
			/**
			 * Moment.js formats.
			 * @link http://momentjs.com
			 */
			timeformat: 'YYYY-MM-DD\\THH:mm:ss.SSSZZ',
			transformer: TextTransformer,
		};
		// tslint:enable:object-literal-key-quotes

		if (options != null) {
			this.setOptions(options);
		}
	}

	/**
	 * Print to console.
	 * NB! This will print to the console based on the log level you have enabled.
	 */
	private print(msg: string, logMask?: number | null): void {
		// return if the log-level is higher than the selected the log-level.
		if (logMask == null) {
			logMask = Mask.LOG;
		}

		if (!matchMask(this._options['log-level'], logMask)) {
			return;
		}

		const entry: LogEntry = {
			group: this.options.group,
			levelNumeric: logMask,
			levelText: this.resolveLogLevel(logMask),
			message: msg,
			timestamp: moment().format(this.options.timeformat),
		};

		// finally printing the output!.
		if (matchMask(logMask, Mask.ERROR)) {
			this._options.outputters.err.call(this, this._options.transformer.call(this, entry));
		} else {
			this._options.outputters.out.call(this, this._options.transformer.call(this, entry));
		}
	}

	/**
	 * Setting a single color in the `ColorMap`
	 */
	private setColor(level: LogLevel | 'TITLE' | 'TIME', color: clc.Format) {
		switch (level) {
			case 'ERROR':
				this.colors[Mask.ERROR] = color;
				break;
			case 'WARN':
				this.colors[Mask.WARN] = color;
				break;
			case 'SUCCESS':
				this.colors[Mask.SUCCESS] = color;
				break;
			case 'LOG':
				this.colors[Mask.LOG] = color;
				break;
			case 'INFO':
				this.colors[Mask.INFO] = color;
				break;
			case 'DEBUG':
				this.colors[Mask.DEBUG] = color;
				break;
			case 'VERBOSE':
				this.colors[Mask.VERBOSE] = color;
				break;
			case 'TITLE':
				this.colors.TITLE = color;
				break;
			case 'TIME':
				this.colors.TIME = color;
				break;
			case 'SILENT':
				break;
			default:
				assertNever(level);
				break;
		}
	}

	/**
	 * Print a debug formatted message.
	 */
	public debug<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (typeof msg === 'string') {
				this.print(msg, Mask.DEBUG);
			} else {
				this.print(this.formatTypes(msg), Mask.DEBUG);
			}
		}
		return firstArg;
	}

	/**
	 * Print an error.
	 */
	public error<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (msg instanceof Error || (typeof msg === 'object' && 'stack' in msg)) {
				this.formatError(msg);
			} else {
				this.print(msg, Mask.ERROR);
			}
		}
		return firstArg;
	}

	/**
	 * Format an error, this is typically used, for string formatting an Exception/Error.
	 */
	public formatError(err: any): void {
		this.print(this.formatTypes(err), Mask.ERROR);
	}

	/**
	 * Format a function call, for the debug level
	 * NB! if passing arguments the function
	 *     every argument, gonna be formatted with the formatTypes() function
	 */
	public formatFunctionCall(functionName: string, args?: any[]) {
		if (args == null) {
			args = [];
		}
		const formattedArgs = [];
		for (const a of Array.from(args)) {
			formattedArgs.push(this.formatTypes(a));
		}
		return this.debug(functionName + `(${formattedArgs.join(', ')})`);
	}

	/**
	 * Format types to string, some types make resively calls.
	 */
	// tslint:disable-next-line:no-reserved-keywords
	public formatTypes(type: any, depth?: number, indent?: number) {
		if (this.options.transformer === JSONTransformer || this.options.transformer === IdentityTransformer) {
			return type;
		}

		// making the proper indentation
		let val;
		if (indent == null) {
			indent = 0;
		}
		if (depth == null) {
			depth = 3;
		}

		const pad = ' '.repeat(indent > 0 ? indent - 1 : 0);

		// primitive types
		if (typeof type === 'number' || typeof type === 'boolean' || type == null) {
			return `${pad}${type}`;
		}

		if (typeof type === 'string') {
			return `${pad}'${type}'`;
		}

		// array is formatted as one-liners
		if (Array.isArray(type) && indent < depth) {
			let str = `${pad}[`;
			// tslint:disable-next-line:forin
			for (const key of type.keys()) {
				val = type[key];
				str += ` ${this.formatTypes(val, indent + 1, depth).trim()}`;
				str += ((key as any) as number) < type.length - 1 ? ',' : ' ';
			}
			return `${str}]`;
		}

		// hashes is formatted with indentation for every level
		// of the object, the values of the properties are also resolved.
		if (typeof type === 'object') {
			const cname = type.constructor.name;
			if (cname === 'Error' || type instanceof Error) {
				let str = `${pad}${type.message}`;
				if (type.stack != null) {
					str += `\n${pad}${type.stack}`;
				}
				return str;
			} else if (cname === 'Object' && indent < depth) {
				let str = `${pad}{`;
				for (const key of Object.keys(type)) {
					val = type[key];
					str += `\n${pad}  ${key}: ${this.formatTypes(val, indent + 1, depth).trim()}`;
				}
				if (str.length > pad.length + 1) {
					str += `\n${pad}`;
				}
				return `${str}}`;
			}
		}

		// print the name of a complex type.
		const isFunc = typeof type === 'function';
		if (isFunc) {
			const cname = type.constructor.name;
			if (isFunc && cname !== 'Function') {
				return `${pad}[Function: ${cname}]`;
			}
			return `${pad}${cname}`;
		}

		// print the name of an unhandled type.
		// typically this will return 'object'
		return `${pad}${typeof type}`;
	}

	/**
	 * Print an info formatted message.
	 */
	public info<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (typeof msg === 'string') {
				this.print(msg, Mask.INFO);
			} else {
				this.print(this.formatTypes(msg), Mask.INFO);
			}
		}
		return firstArg;
	}

	/**
	 * Print an default log formatted message.
	 */
	public log<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (typeof msg === 'string') {
				this.print(msg, Mask.LOG);
			} else {
				this.print(this.formatTypes(msg), Mask.LOG);
			}
		}
		return firstArg;
	}

	/**
	 * Resolve a string representation of the log-level.
	 */
	public resolveLogLevel(mask: number) {
		switch (mask) {
			case Mask.ERROR:
				return 'ERROR';
			case Mask.WARN:
				return 'WARN';
			case Mask.SUCCESS:
				return 'SUCCESS';
			case Mask.INFO:
				return 'INFO';
			case Mask.DEBUG:
				return 'DEBUG';
			case Mask.VERBOSE:
				return 'VERBOSE';
			case Mask.LOG:
				return 'LOG';
			default:
				return 'CUSTOM';
		}
	}

	/**
	 * Set new colors.
	 */
	public setColors(colorMap: { [level: string]: clc.Format }) {
		for (const level of Object.keys(colorMap)) {
			const color = colorMap[level];
			const needle = level.toUpperCase();
			if (isKnownColorType(needle)) {
				this.setColor(needle, color);
			}
		}
	}

	/**
	 * Sets the options.
	 */
	public setOptions(options: Partial<LoggerOptionsInput>) {
		if (options.colors != null) {
			this.setColors(options.colors);
		}
		const currentLoglevel = this.options['log-level'];
		const optionKeys = Object.keys(this.options);
		for (const key of optionKeys) {
			const val = (options as any)[key];
			if (val != null) {
				(this.options as any)[key] = val;
			}
		}

		if (!Number.isInteger(this._options['log-level']) || this._options['log-level'] > Level.VERBOSE) {
			this.options['log-level'] = currentLoglevel;
			this.warn(`Invalid log level is trying to be set: ${options['log-level']}, aborting...`);
		}
	}

	/**
	 * Print a success formatted message.
	 */
	public success<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			this.print(msg, Mask.SUCCESS);
		}
		return firstArg;
	}

	/**
	 * Print a title formatted message.
	 */
	public title<T>(msg: T, ...__: any[]): T {
		for (const m of Array.from(arguments)) {
			this.print(m, Mask.INFO);
		}
		return msg;
	}

	/**
	 * Print a verbose formatted message.
	 */
	public verbose<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			this.print(msg, Mask.VERBOSE);
		}
		return firstArg;
	}

	/**
	 * Print a warning.
	 */
	public warn<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			this.print(msg, Mask.WARN);
		}
		return firstArg;
	}
}

export default AnsiLogger;
