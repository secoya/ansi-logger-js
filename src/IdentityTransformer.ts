import { LogEntry, Transformer } from './AnsiLogger';

/**
 * The identity transformer makes no transformation to the log entry
 * and just returns the [LogEntry] as is.
 */
export class IdentityTransformer implements Transformer {
	public format(entry: LogEntry): LogEntry {
		return entry;
	}
}
