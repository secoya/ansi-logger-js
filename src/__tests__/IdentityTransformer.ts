import { resolveLogLevel, AnsiLogger, LogEntry, Mask } from 'ansi-logger/AnsiLogger';
import { IdentityTransformer } from 'ansi-logger/IdentityTransformer';

describe('IdentityTransformer', () => {
	describe('printer', () => {
		test('provided printer is used', () => {
			const err = jest.fn();
			const out = jest.fn();

			const transformer = new IdentityTransformer({
				err: err,
				out: out,
			});

			const logger = new AnsiLogger({
				transformer: transformer,
			});

			logger.error('error');
			logger.info('info');

			expect(err).toHaveBeenCalledTimes(1);
			expect(out).toHaveBeenCalledTimes(1);
			expect(err).toHaveBeenCalledWith(
				expect.objectContaining({
					group: undefined,
					levelNumeric: Mask.ERROR,
					levelText: resolveLogLevel(Mask.ERROR),
					message: 'error',
					timestamp: expect.any(String),
				}),
			);
			expect(out).toHaveBeenCalledWith(
				expect.objectContaining({
					group: undefined,
					levelNumeric: Mask.INFO,
					levelText: resolveLogLevel(Mask.INFO),
					message: 'info',
					timestamp: expect.any(String),
				}),
			);
		});
	});

	describe('format', () => {
		test('returns input as is', () => {
			const err = jest.fn();
			const out = jest.fn();

			const transformer = new IdentityTransformer({
				err: err,
				out: out,
			});

			const message = null;
			const logEntry: LogEntry = {
				group: 'json',
				levelNumeric: Mask.INFO,
				levelText: resolveLogLevel(Mask.INFO),
				message: message,
				timestamp: 'NOW',
			};

			expect(transformer.format(logEntry)).toBe(logEntry);
		});

		test('format complex types', () => {
			const err = jest.fn();
			const out = jest.fn();

			const transformer = new IdentityTransformer({
				err: err,
				out: out,
			});

			const complexMessage = {
				some: 'complex',
				structure: [1, true],
			};

			const logEntry: LogEntry = {
				group: 'json',
				levelNumeric: Mask.INFO,
				levelText: resolveLogLevel(Mask.INFO),
				message: complexMessage,
				timestamp: 'NOW',
			};

			expect(transformer.format(logEntry)).toMatchSnapshot();
		});
	});

	describe('formatTypes', () => {
		test('input is returned as is', () => {
			const complexMessage = {
				some: 'complex',
				structure: [1, true],
			};

			const err = jest.fn();
			const out = jest.fn();

			const transformer = new IdentityTransformer({
				err: err,
				out: out,
			});

			expect(transformer.formatTypes(complexMessage)).toBe(complexMessage);
		});
	});
});
