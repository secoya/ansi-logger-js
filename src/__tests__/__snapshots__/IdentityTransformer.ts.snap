// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentityTransformer format format complex types 1`] = `
Object {
  "group": "json",
  "levelNumeric": 16,
  "levelText": "INFO",
  "message": Object {
    "some": "complex",
    "structure": Array [
      1,
      true,
    ],
  },
  "timestamp": "NOW",
}
`;
